cli/tools/script/interpreter.rs

```diff
@@ -0,0 +1,248 @@
+    match change {
+      EnvChange::SetEnvVar(var) => self.apply_env_var(var),
+      EnvChange::Cd(new_dir) => {
+        self.cwd = new_dir.clone();
+      }
+    }
+  }
+
+  pub fn apply_changes(&mut self, changes: &[EnvChange]) {
+    for change in changes {
+      self.apply_change(change);
+    }
+  }
+}
+
+enum EnvChange {
+  SetEnvVar(EnvVar),
+  Cd(PathBuf),
+}
+
+pub async fn execute(
+  list: SequentialList,
+  env_vars: HashMap<String, String>,
+  cwd: PathBuf,
+  additional_cli_args: Vec<String>,
+) -> Result<i32, AnyError> {
+  assert!(cwd.is_absolute());
+  let list = append_cli_args(list, additional_cli_args)?;
+  let mut state = EnvState { env_vars, cwd };
+  let mut async_handles = Vec::new();
+  let mut final_exit_code = 0;
+  for item in list.items {
+    match item {
+      SequentialItem::Async(item) => {
+        let state = state.clone();
+        async_handles.push(tokio::task::spawn(async move {
+          let result = execute_command(item, state).await;
+          if let ExecuteCommandResult::Continue(_, _, Some(err)) = result {
+            eprintln!("{}", err);
+          }
+        }));
+      }
+      SequentialItem::Command(item) => {
+        match execute_command(item, state.clone()).await {
+          ExecuteCommandResult::Exit => return Ok(0),
+          ExecuteCommandResult::Continue(exit_code, changes, err) => {
+            state.apply_changes(&changes);
+            if let Some(err) = err {
+              eprintln!("{}", err);
+            }
+            // use the final sequential item's exit code
+            final_exit_code = exit_code;
+          }
+        }
+      }
+    }
+  }
+
+  // wait for async commands to complete
+  futures::future::join_all(async_handles).await;
+
+  Ok(final_exit_code)
+}
+
+enum ExecuteCommandResult {
+  Exit,
+  Continue(i32, Vec<EnvChange>, Option<AnyError>),
+}
+
+fn execute_command(
+  item: ShellCommand,
+  mut state: EnvState,
+) -> BoxFuture<'static, ExecuteCommandResult> {
+  // recursive async functions require boxing
+  async move {
+    let mut changes = Vec::new();
+    let body_result: Result<i32, AnyError> = match item.body {
+      ShellCommandBody::EnvVar(var) => {
+        state.apply_env_var(&var);
+        changes.push(EnvChange::SetEnvVar(var));
+        Ok(0)
+      }
+      ShellCommandBody::Command(command) => {
+        if command.args[0] == "cd" {
+          if command.args.len() != 2 {
+            Err(anyhow!("cd is expected to have 1 argument."))
+          } else {
+            // affects the parent state
+            let new_dir = state.cwd.join(&command.args[1]);
+            match fs_util::canonicalize_path(&new_dir) {
+              Ok(new_dir) => {
+                state.cwd = new_dir;
+                Ok(0)
+              }
+              Err(err) => Err(anyhow!(
+                "Could not cd to {}.\n\n{}",
+                new_dir.display(),
+                err
+              )),
+            }
```

Comment:
Seems like this whole block should be a separate function; wdyt?

Reply:
Yup. Already have done this while adding pipe support 🙂